### PR TITLE
Rich Presence

### PR DESCRIPTION
**Pull request by Liam-Directive**
## Rich Presence

This PR must be paired with the following PRs:
 - Drift base: https://github.com/directivegames/drift-base/pull/114
 - Perseus: https://github.com/directivegames/perseus/pull/90

In this PR, I hook up basic support for the `richpresence.item` endpoint, and not much else. The actual user for these endpoints will be FriendSVC (see Perseus PR).

### CODE REVIEW

File: Drift/Private/DriftBase.cpp [hunk 1]
```diff
@@ -4639,6 +4644,59 @@ void FDriftBase::GetUserIdentitiesByName(const FString& name, const FDriftGetUse
...
+
+        RichPresenceCache.Add(FriendId, Result);
+        (void) Delegate.ExecuteIfBound(true, Result);
+    });
+
+    Request->OnError.BindLambda([this, Delegate, Url](ResponseContext& Context)
+    {
+        FString Error;
+        Context.errorHandled = GetResponseError(Context, Error);
+        DRIFT_LOG(Base, Error, TEXT("Failed to get richpresence from url: %s. Error: '%s'"), *Url, *Error);
```
Comment (dg-long) [inline Drift/Private/DriftBase.cpp]: I'd suggest using `IErrorReporter::Get()->AddError(...)` here to make sure this issue would pop up on Sentry.

See this slack message for info
https://directive.slack.com/archives/C02BZ0SQD/p1718808451847629

File: Drift/Private/DriftBase.cpp [hunk 2]
```diff
@@ -4639,6 +4644,59 @@ void FDriftBase::GetUserIdentitiesByName(const FString& name, const FDriftGetUse
...
+    GetFriendsList(DriftFriends);
+    const int32 NumFriends = DriftFriends.Num();
+    int32 Results = 0;
+
+    for (auto& DriftFriend : DriftFriends)
+    {
+        GetFriendRichPresence(DriftFriend.playerID, FDriftGetFriendRichPresenceDelegate::CreateLambda([&Results, NumFriends, Delegate](bool Success, const FRichPresenceResult& Result)
+        {
+            Results += 1;
+            if (Results == NumFriends)
```
Comment (dg-long) [inline Drift/Private/DriftBase.cpp:4705]: I might be missing something here, but is there a reason why the amount of friends being equal to the amount from `DriftFriends` is necessary for the caching process to be considered a success?

Comment (dg-long) [reply Drift/Private/DriftBase.cpp:4705]: Also what about in the case this isn't a success? The delegate is currently not called in that case (unless that's intended)

Comment (Liam-Directive) [reply Drift/Private/DriftBase.cpp:4705]: It's a way of collating lots of small requests. Each request to drift-base is getting rich presence information for 1 friend.

So for example if you have 10 friends, you only want to trigger the outer delegate once all 10 rich presences have been requested. There isn't another good way to "await" a series of network calls like this. 

If we would be using `TFutures` or `Coroutines` or something than you could do something like a `PromiseAll(...)` implementation, but we're not.

Comment (dg-long) [reply Drift/Private/DriftBase.cpp:4705]: Alright nevermind then

File: Drift/Private/DriftBase.cpp [hunk 3]
```diff
@@ -4639,6 +4654,62 @@ void FDriftBase::GetUserIdentitiesByName(const FString& name, const FDriftGetUse
...
+void FDriftBase::CacheFriendsRichPresence(const FDriftGetFriendsRichPresenceDelegate& Delegate)
+{
+    TArray<FDriftFriend> DriftFriends;
+    GetFriendsList(DriftFriends);
+    const int32 NumFriends = DriftFriends.Num();
+    int32 Results = 0;
+
+    for (auto& DriftFriend : DriftFriends)
+    {
+        CacheFriendRichPresence(DriftFriend.playerID, FDriftGetFriendRichPresenceDelegate::CreateLambda([&Results, NumFriends, Delegate](bool Success, const FRichPresenceResult& Result)
```
Comment (y4ure) [inline Drift/Private/DriftBase.cpp:4701]: There's an issue here, the lambda is capturing the local variable `Results` by reference but CacheFriendRichPresence is async

**Pull request merged**